vault/identity_store_oidc.go

```diff
@@ -556,6 +575,34 @@ func (i *IdentityStore) pathOIDCReadKey(ctx context.Context, req *logical.Reques
 	}, nil
 }
 
+// getRolesReferencingTargetKeyName returns a slice of roles referenced by targetKeyName.
+// Note: this is not threadsafe. It is to be called with Lock already held.
+func (i *IdentityStore) getRolesReferencingTargetKeyName(ctx context.Context, req *logical.Request, targetKeyName string) ([]role, error) {
+	roleNames, err := req.Storage.List(ctx, roleConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var tempRole role
+	rolesReferencingTargetKeyName := make([]role, 0)
+	for _, roleName := range roleNames {
+		entry, err := req.Storage.Get(ctx, roleConfigPath+roleName)
```

Comment:
This seems like an expensive operation to perform on every update. This cleanup operation needs to be done once and then all updates moving forward would play by the rules so I wonder if we have examples of doing this migration in other engines. cc @calvn

Reply:
It can definitely be an expensive operation, but I don't think that key create happens that often. In cases where this is a hot path we'd have a layer of caching but I'm not entirely convinced that we need that in this case.

Reply:
I guess we aren't actually doing any clean ups, just a checking everything is valid and bailing if not. Still iterating over an unknown amount of roles on an update seems expensive.

Reply:
Ya, I understand that listing roles with each key update isn't ideal. I agree with @calvn that it's probably okay in this case. I don't imagine that key updates are frequent. We could eventually add caching if someone notices/reports an intolerable response time.